File: js/combo-system.js
// combo-system.js — Combo 連殺系統
(function() {
  window.SG = window.SG || {};

  var COMBO_TIMEOUT = 2; // 2秒內未殺敵則歸零
  var COMBO_XP_10 = 1.5;
  var COMBO_XP_20 = 2.0;

  function ComboSystem() {
    this.count = 0;
    this.timer = 0;
    this.best = 0;
    this._flash = 0;
  }

  ComboSystem.prototype.addKill = function() {
    this.count++;
    this.timer = COMBO_TIMEOUT;
    if (this.count > this.best) this.best = this.count;
    if (this.count === 10 || this.count === 20 || this.count === 50) this._flash = 0.5;
  };

  ComboSystem.prototype.update = function(dt) {
    if (this.timer > 0) {
      this.timer -= dt;
      if (this.timer <= 0) this.count = 0;
    }
    if (this._flash > 0) this._flash -= dt;
  };

  ComboSystem.prototype.getXPMultiplier = function() {
    if (this.count >= 20) return COMBO_XP_20;
    if (this.count >= 10) return COMBO_XP_10;
    return 1;
  };

  ComboSystem.prototype.getVisual = function() {
    if (this.count < 2) return null;
    return { count: this.count, flash: this._flash > 0 };
  };

  SG.ComboSystem = ComboSystem;
})();

File: unity-project/Assets/Scripts/Progression/ComboSystem.cs
/// <summary>
/// ComboSystem — 連殺系統
/// 對應原始架構: combo-system.js ComboSystem class
/// 追蹤連續擊殺並給予 XP 倍率加成
/// </summary>
public class ComboSystem
{
    private const float DECAY_TIME = 2f;  // 2 秒衰減

    private int _count;
    private float _timer;

    /// <summary>
    /// 增加連殺計數
    /// </summary>
    public void AddKill()
    {
        _count++;
        _timer = DECAY_TIME;
    }

    /// <summary>
    /// 更新衰減計時
    /// </summary>
    public void UpdateTimer(float dt)
    {
        if (_count == 0) return;

        _timer -= dt;
        if (_timer <= 0)
        {
            _count = 0;
        }
    }

    /// <summary>
    /// 取得當前 XP 倍率
    /// </summary>
    public float GetMultiplier()
    {
        if (_count >= 20) return 2.0f;
        if (_count >= 10) return 1.5f;
        return 1.0f;
    }

    /// <summary>
    /// 取得當前連殺數
    /// </summary>
    public int GetCount() => _count;

    /// <summary>
    /// 是否達到里程碑（用於 UI 閃爍）
    /// </summary>
    public bool IsMilestone()
    {
        return _count == 10 || _count == 20 || _count == 50;
    }
}
